elephant/spike_train_processing.py

```diff
@@ -0,0 +1,246 @@
+                complexity < `threshold`, i.e. extracts synchronous spikes.
+            Default: 'delete'
+
+        Returns
+        -------
+        list of neo.SpikeTrain
+            List of spiketrains where the spikes with
+            ``complexity >= threshold`` have been deleted/extracted.
+              * If ``in_place`` is True, the returned list is the same as
+                ``self.input_spiketrains``.
+              * If ``in_place`` is False, the returned list is a deepcopy of
+                ``self.input_spiketrains``.
+
+        """
+
+        if not self.annotated:
+            self.annotate_synchrofacts()
+
+        if mode not in ['delete', 'extract']:
+            raise ValueError(str(mode) + ' is not a valid mode. '
+                             "valid modes are ['delete', 'extract']")
+
+        if threshold <= 1:
+            raise ValueError('A deletion threshold <= 1 would result '
+                             'in the deletion of all spikes.')
+
+        if in_place:
+            spiketrain_list = self.input_spiketrains
+        else:
+            spiketrain_list = deepcopy(self.input_spiketrains)
+
+        for idx, st in enumerate(spiketrain_list):
+            mask = st.array_annotations['complexity'] < threshold
+            if mode == 'extract':
+                mask = np.invert(mask)
+            new_st = st[mask]
+            spiketrain_list[idx] = new_st
+            if in_place:
+                unit = st.unit
+                segment = st.segment
+                if unit is not None:
+                    unit.spiketrains[
+                        self._get_index(unit.spiketrains, st)
+                                     ] = new_st
+                if segment is not None:
+                    segment.spiketrains[
+                        self._get_index(segment.spiketrains, st)
+                                        ] = new_st
+
+        return spiketrain_list
+
+    def annotate_synchrofacts(self):
+        """
+        Annotate the complexity of each spike in the array_annotations
+        *in-place*.
+        """
+        epoch_complexities = self.epoch.array_annotations['complexity']
+        right_edges = (
+            self.epoch.times.magnitude.flatten()
+            + self.epoch.durations.rescale(
+                self.epoch.times.units).magnitude.flatten()
+        )
+
+        for idx, st in enumerate(self.input_spiketrains):
+
+            # all indices of spikes that are within the half-open intervals
+            # defined by the boundaries
+            # note that every second entry in boundaries is an upper boundary
+            spike_to_epoch_idx = np.searchsorted(
+                right_edges,
+                st.times.rescale(self.epoch.times.units).magnitude.flatten())
+            complexity_per_spike = epoch_complexities[spike_to_epoch_idx]
+
+            st.array_annotate(complexity=complexity_per_spike)
+
+        self.annotated = True
+
+    def _get_index(self, lst, obj):
```

Comment:
no need for this method
it's equivalent to `lst.index(obj)` with catching the ValueError

Reply:
```
import neo
import numpy as np
import quantities as pq

st1 = neo.SpikeTrain(np.arange(5)*pq.s, t_stop=5*pq.s)
st2 = neo.SpikeTrain(np.arange(5)*pq.s, t_stop=5*pq.s)

l = [st1, st2]
l.index(st2)
```
causes
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
So `lst.index(obj)` seems to check for equality rather than object identity.

Reply:
You're right, I didn't know that. It's even written in the python docs 

> Return zero-based index in the list of the first item whose value is equal to x.

Then I see the need for the method.